src/hotspot/share/gc/shenandoah/shenandoahReferenceProcessor.cpp

```diff
@@ -58,17 +59,44 @@ static const char* reference_type_name(ReferenceType type) {
   }
 }
 
+template <typename T>
```

Comment:
This looks similar to the code we have for dropping a reference.
https://github.com/openjdk/shenandoah/blob/2785875de3bf86ab20779c5a24ec7e0f69e452aa/src/hotspot/share/gc/shenandoah/shenandoahReferenceProcessor.cpp#L397

Could we use this function there?

Reply:
Let me look into this a bit more closely, in regards to how the card-scanning code treats the traversal of weak pointers.

In other words, it must either be the case that the card-mark done in the code you pointed out above (in `drop`) is not needed or, if it is needed, then at least my comment (and attendant assertion at lines 81 & 82) below is wrong and needs to be suitably adjusted and retested:
```
        Those cards will continue to
    // remain dirty on account of this cross-generational pointer to the referent.
```

I'll update after looking into this more closely. (I'll also then see if the code can be shared like you asked.)

Reply:
A visual examination of the card-scanning code reveals that the card-mark in the portion of code you pointed out is redundant because the card-scanning code always dirties the card containing a cross-generational pointer, irrespective of whether it's a weak or strong pointer. I'll work on eliding that unnecessary check and mark in a separate PR as I want to add some assertions and check it carefully.

Thanks for pointing out this code which I had just glossed over until you pointed it out to me above.

The diffs in my original PR stay unchanged and I'll not try to do anything about sharing the code, since I expect that piece of card-marking code in `drop()` to go away in the PR I promised above, for which I'll open a ticket momentarily.

Reply:
I opened https://bugs.openjdk.org/browse/JDK-8305335 to follow up on that.

Reply:
> Maybe that can be done in a separate PR? (I'm not sure the commentary of this method is entirely relevant to that other situation. If it is, feel free to move forward with testing of the change...

Correct; we'll follow up separately on that in a separate PR & ticket, linked i comment above.